### PR TITLE
Add MemoryProfileCallback

### PR DESCRIPTION
**Pull request by ShriyaPalsamudram**
# What does this PR do ?

This callback enables recording a timeline of memory allocations during training. 

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/memory_profiler.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    Args:
+        dir (Optional[str]): Directory to store the memory profile dump
+
+    Example:
+        >>> callback = MemoryProfileCallback(dir="/mem_profile")
+        >>> trainer = Trainer(callbacks=[callback])
+    """
+
+    def __init__(self, dir: str = "/mem_profile"):
```
Comment (hemildesai) [inline nemo/lightning/pytorch/callbacks/memory_profiler.py:27]: Can we add an option to enable [`warn_tensor_cycles`](https://pytorch.org/blog/understanding-gpu-memory-2/) in this profiler as well?

**Comment by maanug-nv** [comment]
Currently, this will dump the snapshot for all ranks. Should we add an option to only save a snapshot from rank 0? or a specifiable rank?

File: nemo/lightning/pytorch/callbacks/memory_profiler.py (hunk 1 above)
Comment (maanug-nv) [reply nemo/lightning/pytorch/callbacks/memory_profiler.py:27]: The blog post recommends adding that to model init. Is it possible to do this via a Callback? Could we just make this a flag for GPTModel?

Comment (ShriyaPalsamudram) [reply nemo/lightning/pytorch/callbacks/memory_profiler.py]: Addressed in latest commit

**Comment by ShriyaPalsamudram** [comment]
> Currently, this will dump the snapshot for all ranks. Should we add an option to only save a snapshot from rank 0? or a specifiable rank?

Addressed in latest commit

File: nemo/lightning/pytorch/callbacks/memory_profiler.py (hunk 1 above)
Comment (hemildesai) [reply nemo/lightning/pytorch/callbacks/memory_profiler.py]: I think we can just do it in the `on_train_start` method.

Comment (ShriyaPalsamudram) [reply nemo/lightning/pytorch/callbacks/memory_profiler.py]: Also might help catch issues before on_train_start

**Review by maanug-nv — COMMENTED** [review]
Callback LGTM. I have no concerns about memory profiling, since this is almost identical to the one callback we've been using. For memory cycles, we should test by introducing an intentional cycle. Also curious if this produces warnings about any cycles not related to our code (eg in PTL).

**Comment by maanug-nv** [comment]
also, can we allow disabling the profiling, so users can only run the cycle detector?

File: nemo/lightning/pytorch/callbacks/memory_profiler.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import os
+
+import torch
+from pytorch_lightning.callbacks.callback import Callback
+from torch.utils.viz._cycles import warn_tensor_cycles
```
Comment (hemildesai) [inline nemo/lightning/pytorch/callbacks/memory_profiler.py:5]: Can we move this import inside the callback to be safe?

**Review by hemildesai — APPROVED** [review]
LGTM, just one minor comment.

File: nemo/lightning/pytorch/callbacks/memory_profiler.py (hunk 2 above)
Comment (ShriyaPalsamudram) [reply nemo/lightning/pytorch/callbacks/memory_profiler.py:5]: Isn't it better to fail early if there is an issue with this import than fail inside the callback?

**Comment by hemildesai** [comment]
Yeah I was just being cautious to avoid side effects of the import but I don't think this import should have any, so it's probably fine.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
